File: lua/base/base/script/packet.lua
require("classlib")

local VAR_UNDEFINED = 0
local VAR_BOOL = 1
local VAR_INT = 2
local VAR_FLOAT = 3
local VAR_STRING = 4

local VAR_INT64 = 9
local VAR_NUMBER = 10

local VAR_BYTE = 11
local VAR_SHORT = 12

class.Packet()
function Packet:__init(...)
    self.args = VarList()
end

function Packet:addId(id)
    self.args:addInt32(id)
end

function Packet:addInt8(val)
    self.args:addInt8(VAR_BYTE)
    self.args:addInt8(val)
end

function Packet:addInt16(val)
    self.args:addInt8(VAR_SHORT)
    self.args:addInt16(val)
end

function Packet:addInt32(val)
    self.args:addInt8(VAR_INT)
    self.args:addInt32(val)
end

function Packet:addInt64(val)
    self.args:addInt8(VAR_INT64)
    self.args:addInt64(val)
end

function Packet:addFloat(val)
    self.args:addInt8(VAR_FLOAT)
    self.args:addFloat(val)
end

function Packet:addDouble(val)
    self.args:addInt8(VAR_DOUBLE)
    self.args:addDouble(val)
end

function Packet:addString(val)
    self.args:addInt8(VAR_STRING)
    self.args:addInt16(string.len(val))
    self.args:addString(val)
end

function Packet:send()
    net.send(self.args)
end

File: lua/base/base/Packet.cpp
#include "Packet.h"

#include "NetImp.h"

Packet::Packet()
{

}

Packet::~Packet()
{

}

void Packet::addId( LUA_NUMBER val )
{
	m_args.add((int32_t)val);
}

void Packet::addInt8( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::BYTE);
	m_args.add((int8_t)val);
}

void Packet::addInt16( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::SHORT);
	m_args.add((int16_t)val);
}

void Packet::addInt32( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::INT);
	m_args.add((int32_t)val);
}

void Packet::addInt64( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::INT64);
	m_args.add((int64_t)val);
}

void Packet::addFloat( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::FLOAT);
	m_args.add((float)val);
}

void Packet::addDouble( LUA_NUMBER val )
{
	m_args.add((int8_t)Var::NUMBER);
	m_args.add((double)val);
}

void Packet::addString( const char *val )
{
	m_args.add((int8_t)Var::STRING);
	m_args.add((int32_t)strlen(val));
	m_args.add((const char *)val);
}

void Packet::send()
{
	GetNetImp()->writePack(m_args);
}
